sympy/sets/fancysets.py

```diff
@@ -483,6 +483,128 @@ def _intersect(self, other):
+            base = imgset.base_set
+            if base == S.Reals:
+                return imgset.lamda(0)
+            try:
+                iterable = iter(base)
+                return imgset.lamda(next(iterable))
+            except TypeError:
+                raise TypeError
+
+        def _attempt_1(self_expr, final_expr, var_self, var_final, base):
+            var_term = self_expr.coeff(var_self)
+            if var_term == 0:
+                raise NotImplementedError
+            if simplify(
+                    final_expr.subs(var_final, var_self) -
+                    self_expr) % var_term == var_term / 2:
+                # (a*pi + pi + expr1) - ( a*pi  + exp1) == abs(a/2)*pi
+                # can be => (a/2)*pi + expr1
+                final_expr = (var_term / 2) * var_self + first_val(self)
+                ans = ImageSet(Lambda(var_self, final_expr), base)
+            elif simplify(
+                    final_expr -
+                    self_expr) % var_term == 0:
+                # (a*pi + expr1) - ( a*pi  + 2*pi + exp1)= 0
+                # can be => a*pi + expr1
+                final_expr = var_term * var_self + first_val(self)
+                ans = ImageSet(Lambda(var_self, final_expr), base)
+            else:
+                ans = None
+            return ans
+
+        if isinstance(other, ImageSet):
+            try:
+                base = other.base_set
+                base_self = self.base_set
+                if base != base_self:
+                    raise NotImplementedError
+                # use one dummy variable n_self
+                var_self = self.lamda.variables
+                if len(var_self) > 1:
+                    # Not Implemented for multiple lambda variables.
+                    return None
+                var_self = var_self[0]
+                var_final = other.lamda.variables
+                if len(var_final) > 1:
+                    # Not Implemented for multiple lambda variables.
+                    return None
+                var_final = var_final[0]
+
+                # Extracting expr
+                lamb_expr = self.lamda.expr
+                if Poly(lamb_expr, var_self).is_linear:
+                    self_expr = lamb_expr
+                else:
+                    # Not Implemented for non linear ImageSet.
+                    return None
+                if base_self.is_superset(base):
+                    # take the superset
+                    base = base_self
+
+                lamb_expr = other.lamda.expr
+                if Poly(lamb_expr, var_final).is_linear:
+                    final_expr = lamb_expr
+                else:
+                    # Not Implemented for non linear ImageSet.
+                    return None
+
+                ans = _attempt_1(self_expr, final_expr,
+                                 var_self, var_final, base)
+                # Add more ways to club `self` and `other`
+                return ans
+
+            except (
```

Comment:
This is an interesting design.
You are [raising a `NotImplementedError` exception](https://github.com/sympy/sympy/pull/11188/files#diff-c1c17bdd28aafd400bf9415c7d048382R564) in the `try` block and catching the raised error in the `except` clause.
Ping @hargup @aktech